### ISSUE TITLE
Panache Hibernate @Filter not applied to joined entities

### ISSUE THREAD

**Issue by josephearl** [quarkusio/quarkus#26237]
### Describe the bug

Given some entities like:

```
@Entity
@FilterDef(name = "isActive", defaultCondition = "active = true")
public class Child {
  private int childId;
  private int parentId;
  private boolean active;
}
```

```
@Entity
public class Parent {
  private int parentId;
  @OneToMany(fetch = FetchType.EAGER)
  @JoinColumn(name = "parentId", updatable = false, insertable = false)
  @Filter(name = "isActive")
  private List<Child> children;
}
```

And a query method with the filter applied in my repository class for `Parent`:

```
private PanacheQuery<Parent> findParentWithActiveChildren(int id) {
  return find("id = :id", Parameters.with("id", id)).filter("isActive");
}
```

I would expect only children with `active = true` are returned, but all children related to the parent are returned, regardless of their active status

### Expected behavior

Only children with `active = true` are returned in the OneToMany association

### Actual behavior

All children associated with the parent are returned, regardless of their active status

### Quarkus version or git rev

2.9.2.Final

**Comment by geoand** [comment]
Do you have a reproducer project for this?

Asking because we have tests in our testsuite that test this exact feature so it would be very useful to have an exact case where it does not work.

**Comment by josephearl** [comment]
Here's a reproducer - https://github.com/josephearl/quarkus-26237

If you run `ParentEntityRepositoryTest` you will see the tests `shouldFindParentWithActiveChildrenStreamToList` and `shouldFindParentWithActiveChildrenStreamToListAndClosingStream` will fail, but `shouldFindParentWithActiveChildrenList` passes

**Comment by geoand** [comment]
Thanks, I'll give it a try

**Comment by geoand** [comment]
Thanks for the reproducer, very useful indeed.

@FroMage @Sanne is there a reason why `stream().list()` does not engage filters of joined entities? 
If this indeed can't be done, then I think we can detect that and throw an exception?

**Comment by loicmathieu** [comment]
I check the code and it seems to be an issue on Hibernate ORM side. The code path for both `PanacheQuery.stream()` and `PanaceQuery.list()` is the same except for that it returns `jpaQuery.getResultStream();` instead of `jpaQuery.getResultList();`.

This should be reproducible without Panache on plain ORM.

**Comment by FroMage** [comment]
So, this is an ORM bug then? @Sanne or @yrodiere WDYT?

**Comment by yrodiere** [comment]
@FroMage Yes and no. As far as I understand it's more of a lazy loading issue, combined with how Panache applies the filters.

Panache enables filters before executing the query, and disables them just after having executed the query.

With `.list()`, children apparently get loaded eagerly during the `.list()` call, resulting in them being filtered (the filter is still active when children are loaded).

With `.stream()`, children get loaded lazily *after* the `.stream()` call (so after Panache disables the filter), resulting in the children *not* being filtered (the filter is no longer active when children are loaded.

We could argue on whether forcing lazy-loading for `.stream()` makes sense, and there I must admit I don't have a good answer. I imagine there must be history behind that decision.

However, I'd like to point out that applying the filters that way (enabling before the query, disabling just after the query) is inherently dangerous: had `children` been a lazy association (which is the default, and generally recommended for `*ToMany` associations), Hibernate ORM / Panache would have skipped filtering of children with `.list()` as well as `.stream()`.

Really, filters are session-bound constructs, and you should enable them on the session, not on the query, because they have an effect on the whole session, not just on the query. And in fact, that's precisely the behavior we're after in this ticket: children are not even mentioned in the query, but we still want filters to apply to them...
Also, personally I would recommend enabling filters for the whole duration of the session, to avoid weird behavior like the one we're witnessing here.

By the way, there is even a warning in the Panache javadoc about this whole thing:

```java
    /**
     * <p>
     * Enables a Hibernate filter during fetching of results for this query. Your filter must be declared
     * with {@link FilterDef} on your entity or package, and enabled with {@link Filter} on your entity.
     * <p>
     * WARNING: setting filters can only be done on the underlying Hibernate {@link Session} and so this
     * will modify the session's filters for the duration of obtaining the results (not while building
     * the query). Enabled filters will be removed from the session afterwards, but no effort is made to
     * preserve filters enabled on the session outside of this API.
     *
     * @param filterName The name of the filter to enable
     * @return this query, modified
     */
    public <T extends Entity> PanacheQuery<T> filter(String filterName);
```

My recommendation would be to deprecate/remove the `filter` method on `PanacheQuery`, because they're more confusing than anything. As a replacement, either we recommend that people set filters on the Session/EntityManager explicitly, or we provide a Panache-specific API to set things on the session (`Panache.enableFilter`/`Panache.disableFilter`?).

**Comment by josephearl** [comment]
Yes found myself the Panache filter does not work with lazy associations- so am manually enabling and disabling the filter in my repository method once I’ve mapped the db entities to my domain models

Agree that maybe it should be deprecated since it’s functionality is somewhat limited

**Comment by geoand** [comment]
@yrodiere is there any change in ORM6 around this?

**Comment by geoand** [comment]
Gotcha, thanks
